--- a/libraries/Cabal/Cabal/src/Distribution/Simple/Setup.hs	2020-02-14 17:28:59.000000000 +0100
+++ b/libraries/Cabal/Cabal/src/Distribution/Simple/Setup.hs	2020-02-21 15:45:10.681298512 +0100
@@ -359,11 +359,11 @@
     configArgs         = [],
     configPrograms_    = Option' (Just (Last' progDb)),
     configHcFlavor     = maybe NoFlag Flag defaultCompilerFlavor,
-    configVanillaLib   = Flag True,
+    configVanillaLib   = Flag False,
     configProfLib      = NoFlag,
     configSharedLib    = NoFlag,
     configStaticLib    = NoFlag,
-    configDynExe       = Flag False,
+    configDynExe       = Flag True,
     configFullyStaticExe = Flag False,
     configProfExe      = NoFlag,
     configProf         = NoFlag,
