--- a/hadrian/src/Flavour.hs	2022-05-27 12:33:31.000000000 +0200
+++ b/hadrian/src/Flavour.hs	2022-06-22 16:19:05.813781530 +0200
@@ -8,6 +8,7 @@
   , splitSections, splitSectionsIf
   , enableThreadSanitizer
   , enableDebugInfo, enableTickyGhc
+  , disableStripping
   , viaLlvmBackend
   , enableProfiledGhc
   , disableDynamicGhcPrograms
@@ -85,6 +86,7 @@
 flavourTransformers = M.fromList
     [ "werror" =: werror
     , "debug_info" =: enableDebugInfo
+    , "disable_stripping" =: disableStripping
     , "ticky_ghc" =: enableTickyGhc
     , "split_sections" =: splitSections
     , "thread_sanitizer" =: enableThreadSanitizer
@@ -154,6 +156,13 @@
     , builder (Cabal Setup) ? arg "--disable-executable-stripping"
     ]
 
+-- | Disable stripping of libraries.
+disableStripping :: Flavour -> Flavour
+disableStripping = addArgs $ notStage0 ? mconcat
+    [ builder (Cabal Setup) ? arg "--disable-library-stripping"
+    , builder (Cabal Setup) ? arg "--disable-executable-stripping"
+    ]
+
 -- | Enable the ticky-ticky profiler in stage2 GHC
 enableTickyGhc :: Flavour -> Flavour
 enableTickyGhc =
