diff --color -ur a/src/Ide/Plugin/Class/Types.hs b/src/Ide/Plugin/Class/Types.hs
--- a/src/Ide/Plugin/Class/Types.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Ide/Plugin/Class/Types.hs	2024-01-09 18:34:41.924079082 +0100
@@ -207,8 +207,9 @@
         (tmrTypechecked -> gblEnv ) <- useMT TypeCheck nfp
         (hscEnv -> hsc) <- useMT GhcSession nfp
         let binds = collectHsBindsBinders $ tcg_binds gblEnv
+        let sigs = traverse bindToSig binds
         (_, maybe [] catMaybes -> instanceBinds) <- liftIO $
-            initTcWithGbl hsc gblEnv ghostSpan $ traverse bindToSig binds
+            initTcWithGbl hsc gblEnv ghostSpan $ liftZonkM sigs
         pure $ InstanceBindTypeSigsResult instanceBinds
     where
         bindToSig id = do
