--- a/haskell-language-server.cabal	2001-09-09 03:46:40.000000000 +0200
+++ b/haskell-language-server.cabal	2023-07-22 13:56:02.230034265 +0200
@@ -272,7 +272,7 @@
     cpp-options: -Dhls_tactic
 
 common hlint
-  if flag(hlint) && impl(ghc < 9.5)
+  if flag(hlint) && (impl(ghc < 9.5) || flag(ignore-plugins-ghc-bounds))
     build-depends: hls-hlint-plugin == 2.0.0.1
     cpp-options: -Dhls_hlint
 
@@ -339,7 +339,7 @@
 -- formatters
 
 common floskell
-  if flag(floskell) && impl(ghc < 9.5)
+  if flag(floskell) && (impl(ghc < 9.5) || flag(ignore-plugins-ghc-bounds))
     build-depends: hls-floskell-plugin == 2.0.0.1
     cpp-options: -Dhls_floskell
 
@@ -349,12 +349,12 @@
     cpp-options: -Dhls_fourmolu
 
 common ormolu
-  if flag(ormolu) && impl(ghc < 9.5)
+  if flag(ormolu) && (impl(ghc < 9.5) || flag(ignore-plugins-ghc-bounds))
     build-depends: hls-ormolu-plugin == 2.0.0.1
     cpp-options: -Dhls_ormolu
 
 common stylishHaskell
-  if flag(stylishHaskell) && impl(ghc < 9.5)
+  if flag(stylishHaskell) && (impl(ghc < 9.5) || flag(ignore-plugins-ghc-bounds))
     build-depends: hls-stylish-haskell-plugin == 2.0.0.1
     cpp-options: -Dhls_stylishHaskell
 
