--- a/haskell-language-server.cabal	2024-03-11 18:58:38.665244036 +0100
+++ b/haskell-language-server.cabal	2024-03-11 19:06:10.884214206 +0100
@@ -131,6 +131,9 @@
     , process-extras
     , text
 
+  if !flag(cabalfmt)
+    buildable: False
+
 test-suite hls-cabal-fmt-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -203,6 +206,9 @@
     , containers
   hs-source-dirs:   plugins/hls-cabal-plugin/src
 
+  if !flag(cabal)
+    buildable: False
+
 test-suite hls-cabal-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -271,6 +277,9 @@
     DataKinds
     OverloadedStrings
 
+  if !flag(class)
+    buildable: False
+
 test-suite hls-class-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -325,6 +334,9 @@
 
   default-extensions: DataKinds
 
+  if !flag(callHierarchy)
+    buildable: False
+
 test-suite hls-call-hierarchy-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -404,6 +416,9 @@
   default-extensions:
     DataKinds
 
+  if !flag(eval)
+    buildable: False
+
 test-suite hls-eval-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -460,6 +475,9 @@
   default-extensions:
     DataKinds
 
+  if !flag(importLens)
+    buildable: False
+
 test-suite hls-explicit-imports-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -513,6 +531,9 @@
     , transformers
     , unordered-containers
 
+  if !flag(rename)
+    buildable: False
+
 
 test-suite hls-rename-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -572,6 +593,9 @@
   default-extensions:
     DataKinds
 
+  if !flag(retrie)
+    buildable: False
+
 test-suite hls-retrie-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -639,6 +663,9 @@
   default-extensions:
     DataKinds
 
+  if !flag(hlint)
+    buildable: False
+
 test-suite hls-hlint-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -699,6 +726,9 @@
     DuplicateRecordFields
     OverloadedStrings
 
+  if !flag(stan)
+    buildable: False
+
 test-suite hls-stan-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   if (impl(ghc > 8.8.1) && impl(ghc <= 9.2.3) || impl(ghc >= 9.4.0) && impl(ghc < 9.10.0))
@@ -750,6 +780,9 @@
     , text
     , transformers
 
+  if !flag(moduleName)
+    buildable: False
+
 
 test-suite hls-module-name-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -792,6 +825,9 @@
     , transformers
     , containers
 
+  if !flag(pragmas)
+    buildable: False
+
 test-suite hls-pragmas-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -849,6 +885,9 @@
   default-extensions:
     DataKinds
 
+  if !flag(splice)
+    buildable: False
+
 test-suite hls-splice-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -901,6 +940,9 @@
     OverloadedStrings
     RecordWildCards
 
+  if !flag(alternateNumberFormat)
+    buildable: False
+
 test-suite hls-alternate-number-format-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -954,6 +996,9 @@
   default-extensions:
     DataKinds
 
+  if !flag(qualifyImportedNames)
+    buildable: False
+
 test-suite hls-qualify-imported-names-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1003,6 +1048,9 @@
     , transformers
     , vector
 
+  if !flag(codeRange)
+    buildable: False
+
 test-suite hls-code-range-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1057,6 +1105,8 @@
     OverloadedStrings
     RecordWildCards
 
+  if !flag(changeTypeSignature)
+    buildable: False
 
 test-suite hls-change-type-signature-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -1111,6 +1161,9 @@
 
   default-extensions: DataKinds
 
+  if !flag(gadt)
+    buildable: False
+
 test-suite hls-gadt-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1154,6 +1207,9 @@
 
   default-extensions: DataKinds
 
+  if !flag(explicitFixity)
+    buildable: False
+
 test-suite hls-explicit-fixity-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1200,6 +1256,9 @@
   if flag(pedantic)
     ghc-options: -Wwarn=incomplete-record-updates
 
+  if !flag(explicitfields)
+    buildable: False
+
 test-suite hls-explicit-record-fields-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1244,6 +1303,9 @@
     , deepseq
   hs-source-dirs:   plugins/hls-overloaded-record-dot-plugin/src
 
+  if !flag(overloadedRecordDot)
+    buildable: False
+
 test-suite hls-overloaded-record-dot-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1284,6 +1346,9 @@
     , mtl
     , text
 
+  if !flag(floskell)
+    buildable: False
+
 
 test-suite hls-floskell-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -1328,6 +1393,9 @@
     , text
     , transformers
 
+  if !flag(fourmolu)
+    buildable: False
+
 
 test-suite hls-fourmolu-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -1377,6 +1445,9 @@
     , text
     , transformers
 
+  if !flag(ormolu)
+    buildable: False
+
 
 test-suite hls-ormolu-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -1425,6 +1496,8 @@
     , stylish-haskell  ^>=0.12 || ^>=0.13 || ^>=0.14.2
     , text
 
+  if !flag(stylishHaskell)
+    buildable: False
 
 test-suite hls-stylish-haskell-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
@@ -1508,6 +1581,9 @@
     , regex-applicative
     , parser-combinators
 
+  if !flag(refactor)
+    buildable: False
+
 test-suite hls-refactor-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
@@ -1591,6 +1667,9 @@
 
   default-extensions: DataKinds
 
+  if !flag(semanticTokens)
+    buildable: False
+
 test-suite hls-semantic-tokens-plugin-tests
   import:           defaults, pedantic, test-defaults, warnings
   type:             exitcode-stdio-1.0
