--- a/primitive.cabal	2024-08-31 12:36:52.692966768 +0200
+++ b/primitive.cabal	2024-08-31 12:41:26.710109103 +0200
@@ -55,7 +55,7 @@
   Other-Modules:
         Data.Primitive.Internal.Operations
 
-  Build-Depends: base >= 4.9 && < 4.20
+  Build-Depends: base >= 4.9 && < 4.21
                , deepseq >= 1.1 && < 1.6
                , transformers >= 0.5 && < 0.7
                , template-haskell >= 2.11
