--- a/hie-compat.cabal	2023-05-08 16:52:13.832299845 +0200
+++ b/hie-compat.cabal	2023-05-08 16:52:55.541919230 +0200
@@ -52,5 +52,5 @@
     hs-source-dirs: src-ghc90 src-reexport-ghc9
   if (impl(ghc >= 9.2) && impl(ghc < 9.3))
     hs-source-dirs: src-ghc92 src-reexport-ghc9
-  if (impl(ghc >= 9.4) && impl(ghc < 9.5))
+  if (impl(ghc >= 9.4) && impl(ghc < 9.7))
     hs-source-dirs: src-reexport-ghc92
