--- a/src/Ide/Plugin/Floskell.hs	2024-01-09 18:39:09.225284558 +0100
+++ b/src/Ide/Plugin/Floskell.hs	2024-01-09 18:39:35.503613884 +0100
@@ -36,10 +36,10 @@
     let (range, selectedContents) = case typ of
           FormatText    -> (fullRange contents, contents)
           FormatRange r -> (normalize r, extractTextInRange (extendToFullLines r) contents)
-        result = reformat config (Just file) . TL.encodeUtf8 $ TL.fromStrict selectedContents
+        result = reformat config (Just file) $ TL.fromStrict selectedContents
     case result of
       Left  err -> throwError $ PluginInternalError $ T.pack $ "floskellCmd: " ++ err
-      Right new -> pure $ InL [TextEdit range . TL.toStrict $ TL.decodeUtf8 new]
+      Right new -> pure $ InL [TextEdit range $ TL.toStrict new]
 
 -- | Find Floskell Config, user and system wide or provides a default style.
 -- Every directory of the filepath will be searched to find a user configuration.
