--- a/hls-ormolu-plugin.cabal	2023-02-13 17:27:53.609771525 +0100
+++ b/hls-ormolu-plugin.cabal	2023-02-13 17:28:02.272880599 +0100
@@ -23,8 +23,6 @@
     location: https://github.com/haskell/haskell-language-server.git
 
 library
-  if impl(ghc >= 9.5)
-    buildable: False
   exposed-modules:  Ide.Plugin.Ormolu
   hs-source-dirs:   src
   build-depends:
--- a/src/Ide/Plugin/Ormolu.hs
+++ b/src/Ide/Plugin/Ormolu.hs
@@ -56,9 +56,14 @@ provider ideState typ contents fp _ = withIndefiniteProgress title Cancellable $
         CabalNotFound                -> Nothing
         CabalDidNotMention cabalInfo -> Just cabalInfo
         CabalFound cabalInfo         -> Just cabalInfo
+#if MIN_VERSION_ormolu(0,7,0)
+      (fixityOverrides, moduleReexports) <- getDotOrmoluForSourceFile fp'
+      let conf' = refineConfig ModuleSource cabalInfo (Just fixityOverrides) (Just moduleReexports) conf
+#else
       fixityOverrides <- traverse getFixityOverridesForSourceFile cabalInfo
       let conf' = refineConfig ModuleSource cabalInfo fixityOverrides conf
-          cont' = cont
+#endif
+      let cont' = cont
 #else
       let conf' = conf
           cont' = T.unpack cont
