--- a/hls-ormolu-plugin.cabal	2023-02-13 17:27:53.609771525 +0100
+++ b/hls-ormolu-plugin.cabal	2023-02-13 17:28:02.272880599 +0100
@@ -21,10 +21,6 @@
     location: https://github.com/haskell/haskell-language-server.git
 
 library
-  if impl(ghc >= 9.3)
-    buildable: False
-  else
-    buildable: True
   exposed-modules:  Ide.Plugin.Ormolu
   hs-source-dirs:   src
   build-depends:
--- a/src/Ide/Plugin/Ormolu.hs	2023-02-13 17:28:30.370231338 +0100
+++ b/src/Ide/Plugin/Ormolu.hs	2023-02-13 17:29:13.990767243 +0100
@@ -45,7 +45,7 @@
     mkConf o region = defaultConfig { cfgDynOptions = o, cfgRegion = region }
     fmt :: T.Text -> Config RegionIndices -> IO (Either OrmoluException T.Text)
     fmt cont conf =
-      try @OrmoluException $ ormolu conf (fromNormalizedFilePath fp) $ T.unpack cont
+      try @OrmoluException $ ormolu conf (fromNormalizedFilePath fp) cont
 
   case typ of
     FormatText -> ret <$> fmt contents (mkConf fileOpts fullRegion)
