--- a/ormolu.cabal	2022-09-12 12:28:44.275500737 +0200
+++ b/ormolu.cabal	2022-09-12 12:29:23.946045417 +0200
@@ -39,6 +39,13 @@
 
     manual:      True
 
+flag ghc-lib
+    description:
+        Force dependency on ghc-lib-parser even if GHC API in the ghc package is supported
+
+    default:     False
+    manual:      True
+
 library
     exposed-modules:
         Ormolu
@@ -108,12 +115,19 @@
         dlist >=0.8 && <2.0,
         file-embed >=0.0.15 && <0.1,
         filepath >=1.2 && <1.5,
-        ghc-lib-parser >=9.4 && <9.5,
         megaparsec >=9.0,
         mtl >=2.0 && <3.0,
         syb >=0.7 && <0.8,
         text >=2.0 && <3.0
 
+    if ((!flag(ghc-lib) && impl(ghc >=9.4)) && impl(ghc <9.5))
+        build-depends:
+            ghc ==9.4.*,
+            ghc-boot-th -any,
+            ghc-boot -any
+    else
+        build-depends: ghc-lib-parser ==9.4.*
+
     if flag(dev)
         ghc-options:
             -Wall -Werror -Wcompat -Wincomplete-record-updates
