diff --git a/src/Ide/PluginUtils.hs b/src/Ide/PluginUtils.hs
index c5609065c3..e34d19f8b0 100644
--- a/src/Ide/PluginUtils.hs
+++ b/src/Ide/PluginUtils.hs
@@ -28,6 +28,7 @@ module Ide.PluginUtils
     allLspCmdIds',
     installSigUsr1Handler,
     subRange,
+    rangesOverlap,
     positionInRange,
     usePropertyLsp,
     -- * Escape
@@ -277,6 +278,21 @@ fullRange s = Range startPos endPos
 subRange :: Range -> Range -> Bool
 subRange = isSubrangeOf
 
+
+-- | Check whether the two 'Range's overlap in any way.
+--
+-- >>> rangesOverlap (mkRange 1 0 1 4) (mkRange 1 2 1 5)
+-- True
+-- >>> rangesOverlap (mkRange 1 2 1 5) (mkRange 1 0 1 4)
+-- True
+-- >>> rangesOverlap (mkRange 1 0 1 6) (mkRange 1 2 1 4)
+-- True
+-- >>> rangesOverlap (mkRange 1 2 1 4) (mkRange 1 0 1 6)
+-- True
+rangesOverlap :: Range -> Range -> Bool
+rangesOverlap r1 r2 =
+  r1 ^. L.start <= r2 ^. L.end && r2 ^. L.start <= r1 ^. L.end
+
 -- ---------------------------------------------------------------------
 
 allLspCmdIds' :: T.Text -> IdePlugins ideState -> [T.Text]
