--- a/src/Stan/Ghc/Compat.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Stan/Ghc/Compat.hs	2025-12-24 15:14:19.167709260 +0100
@@ -10,12 +10,6 @@
 import Stan.Ghc.Compat902 as Compat
 #elif __GLASGOW_HASKELL__ == 904
 import Stan.Ghc.Compat902 as Compat
-#elif __GLASGOW_HASKELL__ == 906
-import Stan.Ghc.Compat906 as Compat
-#elif __GLASGOW_HASKELL__ == 908
-import Stan.Ghc.Compat906 as Compat
-#elif __GLASGOW_HASKELL__ == 910
-import Stan.Ghc.Compat906 as Compat
-#elif __GLASGOW_HASKELL__ == 912
+#elif __GLASGOW_HASKELL__ >= 906
 import Stan.Ghc.Compat906 as Compat
 #endif
--- a/src/Stan/Ghc/Compat906.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Stan/Ghc/Compat906.hs	2025-12-24 15:14:19.167709260 +0100
@@ -4,7 +4,7 @@
 {-# OPTIONS_GHC -Wno-orphans #-}
 
 module Stan.Ghc.Compat906
-#if __GLASGOW_HASKELL__ == 906 || __GLASGOW_HASKELL__ == 908 || __GLASGOW_HASKELL__ == 910 || __GLASGOW_HASKELL__ == 912
+#if __GLASGOW_HASKELL__ >= 906
     ( -- * Modules
       Module
     , ModuleName
--- a/src/Stan/Hie/Compat.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Stan/Hie/Compat.hs	2025-12-24 15:14:19.167709260 +0100
@@ -8,14 +8,6 @@
 import Stan.Hie.Compat900 as Compat
 #elif __GLASGOW_HASKELL__ == 902
 import Stan.Hie.Compat902 as Compat
-#elif __GLASGOW_HASKELL__ == 904
-import Stan.Hie.Compat904 as Compat
-#elif __GLASGOW_HASKELL__ == 906
-import Stan.Hie.Compat904 as Compat
-#elif __GLASGOW_HASKELL__ == 908
-import Stan.Hie.Compat904 as Compat
-#elif __GLASGOW_HASKELL__ == 910
-import Stan.Hie.Compat904 as Compat
-#elif __GLASGOW_HASKELL__ == 912
+#elif __GLASGOW_HASKELL__ >= 904
 import Stan.Hie.Compat904 as Compat
 #endif
--- a/src/Stan/Hie/Compat904.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Stan/Hie/Compat904.hs	2025-12-24 15:14:19.167709260 +0100
@@ -1,7 +1,7 @@
 {-# LANGUAGE CPP #-}
 
 module Stan.Hie.Compat904
-#if __GLASGOW_HASKELL__ == 904 || __GLASGOW_HASKELL__ == 906 || __GLASGOW_HASKELL__ == 908 || __GLASGOW_HASKELL__== 910 || __GLASGOW_HASKELL__== 912
+#if __GLASGOW_HASKELL__ >= 904
     ( -- * Main HIE types
       ContextInfo (..)
     , HieArgs (..)
--- a/src/Stan/Hie/Debug.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Stan/Hie/Debug.hs	2025-12-24 15:14:19.167709260 +0100
@@ -12,10 +12,6 @@
 import Stan.Hie.Debug902 as Compat
 #elif __GLASGOW_HASKELL__ == 906
 import Stan.Hie.Debug902 as Compat
-#elif __GLASGOW_HASKELL__ == 908
-import Stan.Hie.Debug908 as Compat
-#elif __GLASGOW_HASKELL__ == 910
-import Stan.Hie.Debug908 as Compat
-#elif __GLASGOW_HASKELL__ == 912
+#elif __GLASGOW_HASKELL__ >= 908
 import Stan.Hie.Debug908 as Compat
 #endif
--- a/src/Stan/Hie/Debug908.hs	2001-09-09 03:46:40.000000000 +0200
+++ b/src/Stan/Hie/Debug908.hs	2025-12-24 15:14:19.167709260 +0100
@@ -21,7 +21,7 @@
 -}
 
 module Stan.Hie.Debug908
-#if __GLASGOW_HASKELL__ == 908 || __GLASGOW_HASKELL__ == 910 || __GLASGOW_HASKELL__ == 912
+#if __GLASGOW_HASKELL__ >= 908
     ( debugHieFile
     ) where
 
